OpenRCT2: Implement New Game

Crazycolbster opened the pull request:
Please format your title with what portion of the project this pull request is
targeting and what it's changing.

## What is this fixing or adding?
Adds support for OpenRCT2.

## How was this tested?
Over the past  6 months, we've had several public beta tests, fixing many issues. 

## If this makes graphical changes, please attach screenshots.
![image](https://github.com/ArchipelagoMW/Archipelago/assets/43626579/00d6aeed-e162-4011-9a06-51ed316f075a)
![image](https://github.com/ArchipelagoMW/Archipelago/assets/43626579/994fbe15-f253-434d-9ecb-f8eb9558caa2)
![image](https://github.com/ArchipelagoMW/Archipelago/assets/43626579/9ba4e7fd-f3d4-4bd8-b3c5-69e93ed25658)

Crazycolbster committed:
Man, there's a lot to learn here.

Set up some options. Likely many, many errors

Crazycolbster committed:
Did my previous thing commit? If not,

Created options and some funcitons. Likely many, many errors

Crazycolbster committed:
Made Items and locations usable with some glitches

The program now generates an OpenRCT2 World without logic

Die4Ever committed:
Openrct2Client initial work

Crazycolbster committed:
Minor Changes to OpenRCT2 Items

Crazycolbster committed:
Progress on Networking

Die4Ever committed:
openrct2 proxy fixes

Crazycolbster committed:
Added options for traps and monopoly mode

Crazycolbster committed:
Improved Generation
Fixed Bug where items wouldn't generate

Crazycolbster committed:
Improved the items list
Added new categories to the item list
Improved Options

Crazycolbster committed:
Added massive option to select scenario.
Coming to realise __init__ needs to be redone

Crazycolbster committed:
New Location Algorithm
New items algorithm

Crazycolbster committed:
Fixed error in create_regions

Crazycolbster committed:
Added a single "Beauty Contest" item and location

Crazycolbster committed:
Added weather based filler items
Added options for random scenario from game/expansion pack

Crazycolbster committed:
Added filler items to generator

Crazycolbster committed:
Added park rules to items

Crazycolbster committed:
Created function to find starting ride

Crazycolbster committed:
Added Victory condition. It's useless right now
Started rule generation algorithm

Crazycolbster committed:
Improved prereq code

Crazycolbster committed:
Added category prereqs to prereq rules list

Crazycolbster committed:
Created unlock shop to pass to OpenRCT2

Crazycolbster committed:
Created fill_slot_data and started filling it

Crazycolbster committed:
Added data to be sent to OpenRCT2

Crazycolbster committed:
Added guests as checks

Crazycolbster committed:
Updates to OpenRCT2Client. Fixed connection issues
Bugfixes

Crazycolbster committed:
Client improvements
Visibilty option added

Crazycolbster committed:
Bugfixes

Crazycolbster committed:
Ensured game is linked before connecting to server

Crazycolbster committed:
Fixed bug where packets were dropped while socket was reconnecting

Crazycolbster committed:
Fixed on_package to send game readable PrintJSON's
Fixed categories for location_prices
Fixed bug where lives purchases had hidden prereqs

Crazycolbster committed:
Added support for every scenario

Die4Ever committed:
receive separated packets

Crazycolbster committed:
Fixed generation bug with multiple OpenRCT2 Players

Die4Ever committed:
separated packets fix variable name

Crazycolbster committed:
Fixed generation to be legal again

Crazycolbster committed:
Fix for packets

Crazycolbster committed:
Added preferred intensity as an option

Crazycolbster committed:
Added functionality to random scenario options

Crazycolbster committed:
Added custom title to client

Crazycolbster committed:
Split rules into indiviual settings.
Allow each rule to be permanent or temorary with an item

Crazycolbster committed:
Trap items now correctly identified

Crazycolbster committed:
Fixed generation to not skip items when modulus of 8 = 0
Added shops to rules
Added additional rules to prevent bad balancing

Crazycolbster committed:
Added further rules for generation
If progression rules can be disabled, added rules to sphere 5

Crazycolbster committed:
Fixed Logic Block (Probably)
Reduced default guests to 2000
Limited nausea on prereqs
Reduced default nausea goal

Crazycolbster committed:
Fixed generation bug by moving rules to generate_early

Crazycolbster committed:
Fixed generation bug. This time for reals

Crazycolbster committed:
Added rules for items that have height/landscaping needs

Crazycolbster committed:
Disabled visualize regions since we're done with it

Crazycolbster committed:
Fixed generation issue where list was incorrectly copying

Crazycolbster committed:
Added rules for First Aid Rooms and Cash Machine

Crazycolbster committed:
Should not have pushed without testing. Reverting previous push

Crazycolbster committed:
Added some balancing for Transport and Water Rides

Crazycolbster committed:
Fixed issue with generator and updated ride prereqs

Crazycolbster committed:
Removed any prereqs from the first 3 locations to fix generator bugs with too many rules

Crazycolbster committed:
Balanced nausea on Roller Coaster Category

Crazycolbster committed:
Fixed options to allow disabling objectives

Crazycolbster committed:
Balancing Tweaks

Crazycolbster committed:
Get Linted

Crazycolbster committed:
Setup guide written

Crazycolbster committed:
Added game page (WIP)
Generation tweak

Crazycolbster committed:
Fixed Typo on Unique Ride Requirement

Crazycolbster committed:
Fixed bug where every ride would be required in unique rides if set to 0

Crazycolbster committed:
Added game speed as an optional set of progressive items

Crazycolbster committed:
Updated terminology in options
Made the last 15% of checks always have a prereq

Die4Ever committed:
sockets list

Crazycolbster committed:
Added Connection Pool

Crazycolbster committed:
Uncommented package queue

Crazycolbster committed:
Fixed spelling error

Crazycolbster committed:
Created an .apworld for the game!

Crazycolbster committed:
Fixed issue where multiple cars of a ride causes the same ride to appear twice on unique rides

Crazycolbster committed:
Refactored random scenario selection
Excluded specific scenarios from being selected by the random algorithm

Crazycolbster committed:
Improved random scenario selection

Crazycolbster committed:
Fixed duplicate class names in options

Crazycolbster committed:
Removed option_definitions and replaced with options_dataclass

Crazycolbster committed:
Fixed game-breaking issue with generated files

Crazycolbster committed:
Linting fixes

Crazycolbster committed:
Updated documents

Crazycolbster committed:
Updated .apworld

Crazycolbster committed:
Added "None" option to randomization range

Crazycolbster committed:
Quick fix for a generator breaking typo

Crazycolbster committed:
Fixed Incorrect list of items on Icarus Park

Crazycolbster committed:
Did Past Colby actually fix the issue with random parks this time? Who knows!?

Crazycolbster committed:
Balancing
Added changes to prevent an impossible scenario from being rolled

Crazycolbster committed:
It would appear they don't like me not using self.random. Wacky

Crazycolbster committed:
Cleaned up output code

Crazycolbster committed:
New Option to ignore ride stat changes

Crazycolbster committed:
Added Archipelago Madness Scenario to options
Broke minimum and maximum values for the unlock shop into their own sliders

Crazycolbster committed:
Updated openrct2.apworld

Crazycolbster committed:
Nothing. This added nothing

silasary committed:
Use `select` to better handle reading sockets

Crazycolbster committed:
Fixed bug with time.sleep vs ascynchio.sleep

silasary committed:
Move reading sockets to separate thread, close aborted connections

Crazycolbster committed:
Balanced first 32 checks to ensure beatability. No prereqs on checks 0 through 31

Crazycolbster committed:
Added Alpine Coaster to stat exempt

Crazycolbster committed:
Write the starting ride to the spoiler log now

Crazycolbster committed:
Updated .apworld

Crazycolbster committed:
Limited nausea to a max of 4

Crazycolbster committed:
Added skip options for unlock shop

Crazycolbster committed:
Fixed incorrect items in Crazy Craters

Crazycolbster committed:
Added option to include Cash Machine and First Aid Room to parks that don't have it by default

Crazycolbster committed:
Added Classic Stand-up Roller Coaster for compatibilty with new OpenRCT2 update

qwint reviewed (commented):
not exhaustive, and some of my comments apply to more places in the pr than just commented on
also did not itemize all the notes about cleaning up comments, fixing whitespace etc. to adhere to the pep-8 guidlines (with the exceptions outlined in https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/style.md ), and cleaning up unused imports, etc.

ScipioWright reviewed (commented):
Just some minor comments on something qwint didn't comment on.

Crazycolbster commented on worlds/openrct2/Client.py:
```diff
@@ -0,0 +1,324 @@
+
+
+class OpenRCT2Context(CommonContext):
+    tags = {"AP", "DeathLink"}
```
Huh. I wondered what that was for.

Crazycolbster commented on worlds/openrct2/Client.py:
```diff
@@ -0,0 +1,324 @@
+    tags = {"AP", "DeathLink"}
+    game = "OpenRCT2"
+    items_handling = 0b111  # receive all items for /received
+    want_slot_data = True 
```
The game-side plugin passes all of the communication calls once connected, and the client simply mirrors that. One of the first things the game does when starting is retrieve the list of prices for the unlock shop saved in the slot data

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,576 @@
+    if location_number == 0:
+        return "OpenRCT2_Level_0"
+    elif location_number == 1 or location_number == 2:
+        return "OpenRCT2_Level_0"
```
I think past Colby did this to keep the structure straight in his head, but it can definitely just be combined

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,576 @@
+
+    options_dataclass = openRCT2Options
+    options: openRCT2Options
+    topology_present = True  # show path to required location checks in spoiler
```
Yeah, I thought they might be useful, but that didn't really turn out to be the case

Crazycolbster committed:
Various optimizations/refactorings

Silvris commented on worlds/openrct2/Client.py:
```diff
@@ -0,0 +1,324 @@
+
+
+if __name__ == '__main__':
+    run_as_textclient()
```
This should effectively be split out into a separate file as a root level script. See FactorioClient.py for an example for this.

Crazycolbster replied:
I've been having everyone use the client already, so I'm happy with cutting the code entirely

Crazycolbster commented on worlds/openrct2/Items.py:
```diff
@@ -0,0 +1,94 @@
+    count = 0
+    while count < skips:
+        openRCT2_items.append("Skip")
+        count +=1
```
Oh, so that's how you do that. I knew that at some point in the past, but had totally forgotten

ScipioWright commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,556 @@
+
+
+def get_previous_region_from_OpenRCT2_location(location_number: int):
+    if location_number == 0 or location_number == 1 or location_number == 2:
```
```suggestion
    if location_number in [0, 1, 2]:
```
Or, if there are no location_numbers that are less than 0, you can do
`if location_number <= 2:`

ScipioWright commented on worlds/openrct2/Items.py:
```diff
@@ -0,0 +1,103 @@
+        openRCT2_items.append("Skip")
+        count +=1
+
+    for number, rule in enumerate(item_info["park_rules"]):#Check every rule type
+        if rules[number] == 1:#If it's enabled and can be disabled
+            openRCT2_items.append(rule)#Add an item to disable
```
```suggestion
    for number, rule in enumerate(item_info["park_rules"]):  # Check every rule type
        if rules[number] == 1:  # If it's enabled and can be disabled
            openRCT2_items.append(rule)  # Add an item to disable
```

Feel free to ignore this one if you don't like it like this

ScipioWright reviewed (commented):
Looks like the apworld was committed by accident

Crazycolbster commented on worlds/openrct2/docs/en_OpenRCT2.md:
```diff
@@ -0,0 +1,29 @@
+Roller Coasters: Requires a minimum number of roller coasters to win. Additionally, minimum stats for excitement, intensity, and nausea can be required for extra challenge.
+Pay off loan: Requires the loan to be paid off to win.
+Unique Rides: A new objective. A list of unlockable rides will be given to you. At least 1 of each must be built to win. (Rides count as built once they have posted stats)
+Monopoly Mode: A new objective. (Nearly) Every unowned tile in the park will be purchasable. (Tiles with paths will have construction rights instead) Once you own the entire map, you win! 40 more items will be added to the item pool, 20 land purchase discounts and 20 construction rights discounts. Each one will reduce the cost of their respective tiles by $10, down to the price of free.
```
"New objective" in this case means new as in not an objective in vanilla RCT2.

Crazycolbster committed:
Code optimizations

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,576 @@
+                if candidate not in item_info["non_starters"]:
+                    self.starting_ride = candidate
+                    self.item_frequency[candidate] -= 1
+                    found_starter = True
```
That makes sense

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,576 @@
+            while location < ending_location + 1:
+                locations.append(OpenRCT2Location(self.player, "OpenRCT2_" + str(location),
+                                                  self.location_name_to_id["OpenRCT2_" + str(location)], chosen_region))
+                location += 1
```
Since I'm directly using the count in the locations.append, wouldn't I need to enumerate it?

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,576 @@
+                      "RollerCoasters": [roller_coasters, excitement, intensity, nausea, 0, False],
+                      "RideIncome": [0, False], "ShopIncome": [0, False], "ParkRating": [park_rating, False],
+                      "LoanPaidOff": [pay_off_loan, False], "Monopoly": [monopoly, False],
+                      "UniqueRides": [unique_rides, False]}
```
I'm creating the objects here for the actual plugin to handle. When the game receives this object, it will directly copy the tuples to its memory and track the values. Once the correct value is achieved (Such as getting enough guests in the park), it will flip the boolean. Once all the booleans are flipped, the win signal is sent out. As for the previous list in items.py, the reason is simple. Past Colby is bad at programming and apparently thought he'd use that value someday

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,576 @@
+            count = 0
+            while count != self.item_frequency[item]:
+                logic_table.append(item)
+                count += 1
```
If past Colby could have bothered doing it right the first time, I wouldn't have had to go fix it. That should be better now

Crazycolbster commented on worlds/openrct2/Items.py:
```diff
@@ -0,0 +1,94 @@
+
+    openRCT2_items.append("Beauty Contest")
+    
+    item_table = {name: [base_id + count, True] for count, name in enumerate(openRCT2_items)}
```
Upon looking at the code again (I haven't messed with this in months), it would appear I was doing things really weirdly. I've fixed it to be a single list that future functions all use.

Crazycolbster committed:
Major refactoring

Crazycolbster commented on .gitignore:
I think I removed it now? .apworld is definitely gone now

Crazycolbster commented on worlds/openrct2/docs/commands_en.md:
```diff
@@ -0,0 +1,66 @@
+- `!!setMaxSpeed x` Sets the maximum speed the game will allow, from 1 to 5.
+- `!!resendChecks` Resends all the purchased checks, in case the connector is bad at its job.
+
+#### Remote Commands
```
I simply copy/pasted. I'm happy to get rid of it

Crazycolbster commented on worlds/openrct2/docs/en_OpenRCT2.md:
```diff
@@ -0,0 +1,29 @@
+
+## Quick Links
+- [Setup Guide](../../../../tutorial/OpenRCT2/setup/en)
+- [Settings Page](../../../../games/OpenRCT2/player-settings)
```
Do you know if there's an alternate way to check? Ubuntu seems to throughly disaprove of webhost.py

Crazycolbster committed:
Further refactoring

Die4Ever committed:
client.py cleanup

Crazycolbster committed:
Fixed style in imports

Crazycolbster committed:
Networking updates

Crazycolbster committed:
Client Improvements

Crazycolbster committed:
Fixed random scenarios not generating

Die4Ever committed:
split class OpenRCT2Socket into its own file

Die4Ever committed:
automated tests initial work

Die4Ever committed:
don't run network tests in GitHub Actions

Die4Ever committed:
OpenRCT2 tests fix GitHub Actions import

Die4Ever committed:
OpenRCT2 move tests to xtests because they aren't ready to run yet in GitHub Actions

Die4Ever committed:
OpenRCT2 move xtests to tests

Crazycolbster committed:
Added a working test

Crazycolbster committed:
Added a working test

Die4Ever committed:
openrct2 tests some cleanup

Die4Ever committed:
openrct2 inbound packets queue

Die4Ever committed:
openrct2 fix network test

Crazycolbster committed:
Slight change in import

Crazycolbster committed:
Created test function that fails. Gonna need to fix this

Die4Ever committed:
automated tests fixes

Crazycolbster committed:
Updated documentation

Crazycolbster committed:
Updated Setup Doc to say don't follow setup doc (Until release)

Crazycolbster committed:
Slight balance change for Async games.
Gravity Gardens won't be rolled with the random scenario option

Crazycolbster committed:
Fixed issues with display on the website

Added option groups

Crazycolbster committed:
Moved Scenario Options to top of options page.

Added icon for OpenRCT2 launcher

Crazycolbster committed:
Added code to fetch icon from Github for .apworlds

ScipioWright commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,862 @@
+    option_full = LocationBalancingMode.full.value
+    default = LocationBalancingMode.compromise.value
+
+class SelectedScenario(Choice):
```
You should probably check how this looks on webhost. I'm hoping it looks fine

Crazycolbster replied:
I ran through it pretty recently. I think everything looks good!

Crazycolbster committed:
P*ck it, we're just going to include the picture in the data folder by default

Exempt-Medic commented:
Don't forget to update Readme.md and CODEOWNERS

Exempt-Medic commented:
What are the `.gitignore` changes for?

Exempt-Medic commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,526 @@
+    topology_present = False  # show path to required location checks in spoiler
+    item_name_to_id = {name: id for id, name in enumerate(item_info["all_items"], base_id)}
+    location_name_to_id = {name: id for id, name in enumerate(location_info["all_locations"], base_id)}
+    item_name_groups = {
```
You should probably make these more user-friendly, so "Roller Coasters" instead of "roller_coasters"

Crazycolbster replied:
It took some debugging, but I finally got it

Exempt-Medic commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,862 @@
+        return self.value
+
+
+class LocationBalancing(Choice):
```
Where/How does this option work exactly? I don't see it used anywhere

Crazycolbster replied:
Welp. This is embarassing. When I first copied this over, I conflated "Progression Balancing" with "Location Balancing", and never looked at it again.

Exempt-Medic reviewed (changes requested):
I did not look at `Client.py`, the socket, or the unit test. I looked through the rest of the files, although I'll need to look at `__init__.py` again later, since I got a bit lost. Various comments, I have not done generations yet

Crazycolbster commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,862 @@
+    unlockable = 1
+    on = 2
+
+class OpenRCT2OnToggle(DefaultOnToggle):
+    @property
+    def result(self) -> bool:
+        return self.value
+
+
+class OpenRCT2Toggle(Toggle):
+    @property
+    def result(self) -> bool:
+        return self.value
```
I'll be honest, this whole section was originally based on Overcooked 2's implementation, so I just copy-pasted those.

Jarno458 replied:
I think these can be removed as i dont see you use them

Crazycolbster committed:
Commented out longest test to save time

Crazycolbster commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,862 @@
+    option_full = LocationBalancingMode.full.value
+    default = LocationBalancingMode.compromise.value
+
+class SelectedScenario(Choice):
```
This is the order scenarios unlock in the game, so it matches the in game scenario select menu exactly.

Crazycolbster commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,862 @@
+
+    "Nothing" tells you nothing about the item you'll purchase. 
+    
+    Recipient tells you who will recieve the item, but not what they'll receive. 
+    
+    Full tells you what you're buying and who recieves it.
```
0 stars to the dude in the 17th century who decided "i before e except after c".

Crazycolbster committed:
Typo Fixes

Crazycolbster commented on worlds/openrct2/Options.py:
A lot of this code is based off the work of the Overcooked 2 world. The options file was orginally a direct copy that I started making changes to, and the IntEnums were there in the first place. They worked pretty well for me!

Crazycolbster committed:
Removed Location Balancing

Fixed typos/typing conventions

Crazycolbster committed:
Refactored item groups code for better user friendlyness.

Fixed typos

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,526 @@
+                    add_rule(region_entrance, lambda state: state.has("Cash Machine", self.player, 1))
+                if "First Aid" in self.item_table:
+                    add_rule(region_entrance, lambda state: state.has("First Aid", self.player, 1))
+            add_rule(region_entrance, lambda state: state.has_group("rides", self.player, num_rides))
```
Duplicates are fine. All rides are listed as progression instead of filler, and this is to help push progression items up front to keep early sphere items from appearing super late in the game.

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,526 @@
+                         lambda state, selected_prereq=selected_item: state.has(selected_prereq, self.player))
+                # Only add rules if there's an item to be unlocked in the first place
+                if (selected_item in item_info["requires_height"]) and (
+                        self.options.forbid_high_construction.value == 1):
```
When I first wrote this, I didn't realize that was an option. I think I've got them all now

Crazycolbster committed:
Improved code readability

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,526 @@
+        pay_off_loan = self.options.pay_off_loan.value
+        monopoly = self.options.monopoly_mode.value
+        unique_rides = self.unique_rides
+        objectives = {"Guests": [guests, False], "ParkValue": [park_value, False],
```
This is being passed directly to the plugin in OpenRCT2, which tracks progress on the objectives with it. When enough guests are reached (or any other objective is complete), the value will flip, and when all are true, the game is won

Crazycolbster committed:
Bug fixes, implementing suggestions from people better at coding than me

Crazycolbster committed:
Added get_filler_item_name.
Added ensurances that categories of required rides will be possible to build.
If Allow High Construction is disabled, automatically remove items that can not be built at all.
Started list of parks without water access.
Cleaned code.

Crazycolbster committed:
Forgot to comment out chatty code before

Crazycolbster commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,526 @@
+            add_rule(self.multiworld.get_region("Victory", self.player).entrances[0],
+                     lambda state, selected_prereq=ride: state.has(selected_prereq, self.player))
+
+    def generate_basic(self) -> None:
```
To my understanding, I want all of the requirements to be set beforehand for the victory path in the spoiler log. Is there an issue with it being this late?

Exempt-Medic replied:
You could do it at the end of `set_rules`. I don't think there's any issue with where it is, just a bit unusual

Crazycolbster commented on worlds/openrct2/data/scenarios.json:
```diff
@@ -0,0 +1,454 @@
+        "mineral_park",
+        "coaster_crazy",
+        "urban_park",
+        "geoffery_gardens",
```
Yep! It's a weird spelling in the game.

Crazycolbster committed:
Updated park names to reflect RCT2 expansions change in OpenRCT2

Crazycolbster committed:
Updated tests.
Style updates.

Crazycolbster committed:
Added rides from new update

Crazycolbster committed:
Changed locations from Open_RCT2_X to a colored branch system

Crazycolbster committed:
Added option for every item in any scenario

Crazycolbster committed:
Fixed rare issue where stalls would occasionally fill up every slot

Crazycolbster committed:
Added option to have unlock shop identify what category an item is

massimilianodelliubaldini commented on .gitignore:
```diff
@@ -200,3 +200,28 @@ minecraft_versions.json
 .LSOverride
 Thumbs.db
 [Dd]esktop.ini
+worlds/sm_map_rando.apworld
```
I really don't think you should be adding all these to the project-level gitignore. You can add these to your local gitignore of course, but you should leave the change unstaged so it doesn't make it into a commit. You can also tell a file to skip the git worktree so git thinks no change occurred to it in the first place.

Crazycolbster replied:
Now I just need to figure out how to make a local .gitignore

massimilianodelliubaldini commented on worlds/openrct2/Client.py:
```diff
@@ -0,0 +1,117 @@
+        print(args)
+        self.gamesock.sendobj(args)
+
+    # async def disconnect(self, allow_autoreconnect: bool = False):
```
Do you still need these comment functions? If so, maybe add a comment to remind yourself (and others). Otherwise, best to remove them.

massimilianodelliubaldini commented on worlds/openrct2/Client.py:
```diff
@@ -0,0 +1,117 @@
+        self.ui = TextManager(self)
+        self.ui_task = asyncio.create_task(self.ui.async_run(), name="UI")
+
+# Replacing this with code from Serpent.ai to make a .apworld
```
If the code is outright replaced by the `main` function below, you can remove this comment block as well.

massimilianodelliubaldini commented on worlds/openrct2/OpenRCT2Socket/__init__.py:
```diff
@@ -0,0 +1,204 @@
+                except Exception as e:
+                    logex("Error in connect game: ", e)
+        except socket.timeout as e:
+            #print('error connecting to game', e)
```
I guess these commented prints were for debugging. Should you be logging these exceptions instead? Catching an exception without any handling is kind of a code smell.

Crazycolbster replied:
This code was written by @Die4Ever, and I'm scared about breaking something if I mess with it. If I recall correctly though, we actually see that one a lot due to the unique way OpenRCT2 handles sockets, though I may be remembering wrong on that front.

Crazycolbster replied:
If I understand correctly, the purpose of catching these exceptions is to inform the client that we indeed are expecting them and it isn't a problem. Don't crash the program over these. I'm going to go ahead and write that in the code as an explicit comment.

massimilianodelliubaldini commented on worlds/openrct2/Client.py:
```diff
@@ -0,0 +1,117 @@
+    def on_package(self, cmd: str, args: dict):
+        # if cmd == "Connected":
+        #     self.game = self.game#slot_info[self.slot].game
+        print("PACKAGE!!!")
```
I admire the enthusiasm but, maybe provide more info on the print? The `cmd` for instance? You print the args later anyway.

massimilianodelliubaldini commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,545 @@
+                            nausea = 0
+                            # 3 coin flips to determine what, if any, stat prereqs will be used
+                            if self.random.random() < .5:
+                                excitement = round(self.random.uniform(self.options.shop_minimum_excitement.value, self.options.shop_maximum_excitement.value))
```
Lines that might be too long (1/n)

massimilianodelliubaldini commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,545 @@
+                        while not category_selected:
+                            category = self.random.choice(item_info["ride_types"])
+                            for ride in possible_prereqs:
+                                if ride not in item_info["requires_landscaping"]: # Too many parks are unpredictable with water access, especially if landscaping is disabled
```
Lines that might be too long (2/n)

massimilianodelliubaldini commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,545 @@
+                                        category_selected = True    #e.g. thrill rides won't be required if none can be unlocked at that point
+                        set_openRCT2_rule("category", category, number)
+                        # print("Added requirement for: " + category)
+                        if category == "Roller Coasters" and any(item in possible_prereqs and item not in item_info["stat_exempt_roller_coasters"] for item in possible_prereqs):
```
Lines that might be too long (3/n)

massimilianodelliubaldini commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,545 @@
+                            nausea = 0
+                            # 3 coin flips to determine what, if any, stat prereqs will be used
+                            if self.random.random() < .5:
+                                excitement = round(self.random.uniform(self.options.shop_minimum_excitement.value, self.options.shop_maximum_excitement.value))
```
Lines that might be too long (4/4)

massimilianodelliubaldini reviewed (changes requested):
Copying Violet's comment from another open game PR:

> Hi. We are trying to get rid of & be more strict about loading data from text files via file-io because it cannot be optimised in the binary builds.
>
> Do you think it'd be possible to rewrite the json data in base Python? Or, what some people do is they write a conversion script from .json to .py, so that they can still keep the .json as the source of truth.
>
> Hopefully it shouldn't be too difficult, in a lot of cases json is already almost valid Python syntax.

massimilianodelliubaldini commented:
I did a look through on the code, no testing. Mostly commented on readability, code cleanliness, etc.

Crazycolbster committed:
Added options for ride length and minimum ridership in the in-game shop

Crazycolbster committed:
Removed all .json's from the apworld for better binary builds. Or so I've been told. I'm just a simple Electrical Engineer

Crazycolbster committed:
Style fixes

Crazycolbster committed:
Comment Trimming

Crazycolbster committed:
Update .gitignore

Crazycolbster commented on worlds/openrct2/Options.py:
```diff
@@ -625,30 +633,65 @@ class ShopMinimumNausea(Range):
     default = 4
 
+class ShopMaximumLength(Range):
+    """If the shop determines you need a ride with a miniimum length, this value will be the highest it can ask for.
```
It's weird to be sure. Basically, the generator is going to pick a value between the minimum and maximum, and in game, the player will have to build a ride that exceeds whatever that value is, thus the weirdly placed "minimum"

Crazycolbster committed:
Update worlds/openrct2/Options.py

Co-authored-by: massimilianodelliubaldini <8584296+massimilianodelliubaldini@users.noreply.github.com>

massimilianodelliubaldini commented:
I just noticed there's no changes to CODEOWNERS or Readme.md. Are you planning on being the world maintainer post-merge?

Crazycolbster commented:
> I just noticed there's no changes to CODEOWNERS or Readme.md. Are you planning on being the world maintainer post-merge?

That is the plan, yes. I thought the readme would be based off the main readme for Archipelago. Are they per individual worlds?

massimilianodelliubaldini commented:
Yeah that's what I'm referring to. As part of this PR, you should add an entry for yourself in the top-level readme and the CODEOWNERS file.

[Take Inscryption as an example.](https://github.com/ArchipelagoMW/Archipelago/pull/3621/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) You would add the game name at the bottom of this list in the readme, and then you would add the game name, path to `/world/game`, and your GitHub username (and any collaborators who agree to be there). The CODEOWNERS file should have games listed alphabetically, so you would add OpenRCT2 somewhere in the O's.

Crazycolbster committed:
Added .apworld version
.apworld version sends to plugin to ensure compatibility

Crazycolbster committed:
Added option to balance the guest requirements for the unlock shop

Crazycolbster committed:
Commented out some prints

Crazycolbster committed:
Added Loan Shark Trap

Crazycolbster committed:
Added OpenRCT2 to README and CODEOWNERS

massimilianodelliubaldini commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,912 @@
+    like Spiral Slides will tend towards the minimum, while high throughput rides like roller coasters will 
+    tend towards the maximum.
+    """
+    display_name = "Randomize Park Values"
```
```suggestion
    display_name = "Balance Guest Counts"
```

Forgot to change the display name. :)

Crazycolbster committed:
Fixed typo in maximum guest option

Co-authored-by: massimilianodelliubaldini <8584296+massimilianodelliubaldini@users.noreply.github.com>

massimilianodelliubaldini reviewed (commented):
From here on, the code looks good to me. I haven't played the game so I will just change my review to Comment-only.

Crazycolbster committed:
Clarified option about length being in meters

massimilianodelliubaldini commented:
You should add a blank \_\_init\_\_.py file to your data subfolder so your import statements still work in frozen builds. See https://github.com/ArchipelagoMW/Archipelago/pull/4773.

Crazycolbster committed:
Added __init__.py to data folder for frozen builds.

massimilianodelliubaldini commented:
I don't mean to keep coming back with something to nitpick, honest. I just wanna share the news as I learn about it myself, and to preempt future headaches. 

For `data/openrct2icon.png`, considering it looks to be OpenRCT2's actual icon, would that be allowed in Archipelago? Is it considered a copyrighted or derivative work, and/or is it safe to include as OpenRCT2 is licensed under GPL3? @LegendaryLinux I don't know if this is something you can help clarify. If it ends up being a problem, I don't mind volunteering to help make a new icon.

massimilianodelliubaldini commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,590 @@
+)
+
+
+LauncherComponents.icon_paths['openrct2icon'] = local_path('data', 'openrct2icon.png')
```
```suggestion
LauncherComponents.icon_paths['openrct2icon'] = f"ap:{__name__}/icons/openrct2icon.png"
```

And then move the icon from the root `data` folder to a `worlds/openrct2/icons` folder. This will take advantage of the newer icon API.

Crazycolbster replied:
You should never apologize for improving a project!

massimilianodelliubaldini reviewed (changes requested):
Another one, sorry...

Crazycolbster committed:
Changed icon code to reflect newer API

Co-authored-by: massimilianodelliubaldini <8584296+massimilianodelliubaldini@users.noreply.github.com>

Crazycolbster committed:
Updated Icon Location
Fixed bug where prereqs had too many tracked rides

Crazycolbster committed:
Fixed broken link in the setup doc

massimilianodelliubaldini reviewed (approved):
Long time coming. Handled all my feedback. Code looks good, I did a static review of everything, including the client and socket code. I didn't test the game but I've seen it played in very large multi world streams with no obvious bugs.

Crazycolbster committed:
Fixed ride lists for Loopy Landscapes scenarios.
Minor readibility changes.

Crazycolbster committed:
Updated Version Number

Crazycolbster committed:
Tweaked the minimum guest count maximums from 800 default, 2000 max to 400 default, 1000 max.

Crazycolbster committed:
Tweaked default and maximum values for length and skips.

Crazycolbster committed:
I changed my mind. I want tons of skips possible if the players prefer it

Crazycolbster committed:
Attempt to seriously discourage the player from changing the preferred intensity setting.

Crazycolbster committed:
Forgot to change the maximum allowed value on minimum total customers.

cjmang committed:
Making it possible to make unique rides local or remote

cjmang committed:
slight cleanup

Crazycolbster committed:
Hypothetically fixed the webhost links.

Hypothetically fixed the random scenarios sometimes picking a disabled scenario.

Crazycolbster committed:
Update worlds/openrct2/__init__.py

Changed .value to string check.

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

Crazycolbster committed:
Added fireworks option.

Added descriptive text for Launcher, including a random second half.

Crazycolbster committed:
Updated version number.

cjmang committed:
Removing the option to specify where in the multiworld unique rides get placed

Crazycolbster committed:
Fixed issue where description text would cause older versions of the Launcher to crash (6 people have died in an accident on Launcher 1!).

Crazycolbster committed:
Merging branches

Jarno458 commented on worlds/openrct2/data/location_info.py:
```diff
@@ -0,0 +1,812 @@
+location_info = {
+    "all_locations":[
```
instead of having to hardcode these by hand i recommend just use a list comphrehension
```
[
location_info = {
    "all_locations":
    f"{color}{i}" 
    for i in range(100)
    for color in ("White", "Black", "Green", "Blue", "Yellow", "Gold", "Silver", "Celadon", "Pink")
]
}
```

Jarno458 commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,932 @@
+    option_build_your_own_six_flags_over_texas = Scenario.build_your_own_six_flags_over_texas.value
+    default = Scenario.archipelago_madness_vanilla.value    
+
+class DeathLink(Choice):
```
You can remove this and use the mixin instead

Crazycolbster replied:
I found a middle ground  where instead of using mixins, I extend the deathlink class to append my own custom text to the message.

Jarno458 commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,932 @@
+]
+
+@dataclass
+class openRCT2Options(PerGameCommonOptions):
```
Now your wondering what is a mixin, well its what you put here 
`class openRCT2Options(PerGameCommonOptions, DeathLinkMixin):`
you can import `DeathLinkMixin` at the top from Options like everything else

Jarno458 commented on worlds/openrct2/Options.py:
```diff
@@ -0,0 +1,932 @@
+    skips: Skips
+
+    # deathlink
+    death_link: DeathLink
```
the mixin takes care of this, so you now also remove this

Jarno458 commented on worlds/openrct2/test/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from test.bases import WorldTestBase
+
+
+class CV64TestBase(WorldTestBase):
```
CastleVania64 Test Base? 
since this seems to be just empty you can remove this or if you need it, rename it to OpenRCT2TestBase but test bases are only needed if you need to do some commen setup across multiple tests, if not you can just use `WorldTestBase`

Jarno458 reviewed (changes requested):
I see lots of #print statements all through your code, this a sign of working in the wrong way, you using debug prints to help you debug issues, instead use proper breakpoints and debug your code by attaching a debugger to Generate.py instead., so you can live inspect how your code behaves

Crazycolbster committed:
Added awards and associated options to generation.

Crazycolbster committed:
Added archipelago.json file

NewSoupVi commented:
Clearly something went very wrong here, the PR is suddenly changing a bunch of core files. Adding waiting on: author until the PR is fixed

Crazycolbster committed:
Updated version number

Crazycolbster committed:
Changed default for awards to all

Crazycolbster committed:
Fixed ride list in Hydro Hills

Crazycolbster commented on worlds/openrct2/Constants.py:
```diff
@@ -0,0 +1,160 @@
+# ID of first item and location, could be hard-coded but code may be easier
+# to read with this as a property.
+
+base_id = 2000000
```
This is true, though there's enough technical debt I'd have to take on on the client/plugin side to change it back to 0 that I'm not going to worry about it

Crazycolbster committed:
Locations now go to 999 for each color instead of 99

Crazycolbster committed:
Altered DeathLink to extend the Deathlink Class.

Crazycolbster committed:
The Unlock Shop now appears in the Spoiler.

Crazycolbster committed:
Added Food Poisioning Traps

Crazycolbster committed:
Minor code improvements from review. Removed old icon that might not have been MIT compliant.

Crazycolbster committed:
Further minor adjustments from code review.

Mysteryem commented on worlds/openrct2/__init__.py:
```diff
@@ -0,0 +1,761 @@
+                    add_rule(region_entrance, lambda state: state.has("Cash Machine", self.player, 1))
+                if "First Aid" in self.item_table:
+                    add_rule(region_entrance, lambda state: state.has("First Aid", self.player, 1))
+            add_rule(region_entrance, lambda state: state.has_group("Rides", self.player, num_rides))
```
There is a lambda capture issue here with `num_rides`. The rules that require `num_rides` rides from the `"Rides"` group currently each end up using whatever value that `num_rides` ends up on when the `while` loop finishes.

```suggestion
            add_rule(region_entrance, lambda state, ride_count=num_rides: state.has_group("Rides", self.player, ride_count))
```

General information on lambda capture issues:
https://docs.python.org/3/faq/programming.html#why-do-lambdas-defined-in-a-loop-with-different-values-all-return-the-same-result

This lambda capture issue in OpenRCT2 was found by running Eijebong's [Archipelago fuzzer](https://github.com/Eijebong/Archipelago-fuzzer/) using a [hook](https://github.com/Mysteryem/Archipelago-fuzzer/blob/mysteryem_hooks/hooks/detect_rule_variable_capture_issues.py) I wrote to detect lambda capture issues.

Crazycolbster replied:
And fixed. Well, fixed by @Eijebong.

Crazycolbster committed:
Fixed incorrect ride list on Magic Quarters.

Crazycolbster committed:
Gentle rides will no longer appear with "Prefer High Intensity" is enabled and Roller Coasters will no longer appear with "Prefer Low Intensity" enabled.

Crazycolbster committed:
All scenarios now include log flume and merry go round to ensure all awards are achievable.

Crazycolbster committed:
Awards are now exclusively filler. P*cking Useful Progression items.

Crazycolbster committed:
Now Preferred Intensity *actually* works as intended. Prefer less intense will limit coasters/thrills to 1 with no other stats and vise versa for More intense.

Crazycolbster committed:
Fixed ANOTHER scenario with the wrong items.

Crazycolbster committed:
Added 2 food stalls to Extraterrestrial Extravaganza to make Best Food award possible.

Crazycolbster committed:
Changed useful filler to regular filler to prevent generaton crash with filler items.

Crazycolbster committed:
Updated Version Number

Eijebong committed:
Fix lambda capture in ride gating rules

The lambda in the level progression loop captured num_rides by
reference, so all regions ended up using the value from the last
iteration (which was more than likely 0 as everything past level 5 had
`num_rides` set to 0, effectively making this dead code. Each level now
correctly captures its own threshold.

Also cap num_rides to the actual number of rides in the pool, since
the hardcoded thresholds can exceed what a scenario provides which would
end up with Victory not being accessible.

Crazycolbster committed:
Fixed Thunder Rock Scenario Items

Crazycolbster committed:
Added tests that may or may not work. Fixed bug where identical rides could appear on the unique rides objective.

Crazycolbster committed:
Merge pull request #21 from Eijebong/openrct2-lambda

Fix lambda capture in ride gating rules

Crazycolbster committed:
Implemented granular control of land/construction rights prices and discounts.

Crazycolbster committed:
Added specific shops/scenery groups over general stalls and scenery.

Crazycolbster committed:
Changed message to not break randomness in the main program.

Crazycolbster committed:
Minor spelling updates.

Crazycolbster committed:
"Poisoning" is now spelleled correctly.

Crazycolbster committed:
Fully implemented Traplink.

Crazycolbster committed:
Changed logic and removed Set() that could break reproducability.

Crazycolbster committed:
Sync changes from upstream main

Crazycolbster committed:
Updated code for clarity, finally using Enums right.

Fixed bug where the option for random Loopy Landscapes and Corkscrew Follies scenarios were flipped.

Fixed bug where Awards the Enum and Awards the class shared the same name.

Crazycolbster committed:
Added precise options to control required rides in the unlock shop.

Crazycolbster committed:
Fixed bug where some parks didn't have enough food stalls to earn award.

Crazycolbster commented on worlds/openrct2/docs/setup_en.md:
```diff
@@ -0,0 +1,79 @@
+
+
+## Quick Links
+- [Main Page](../../../../games/OpenRCT2/info/en)
+- [Options Page](../../../../games/OpenRCT2/player-options)
```
Finally getting around to this one. Sorry about the (year long) delay! Testing on my machine, the pages seem to load fine as is, but applying the shorter path then causes a 404 error. I'm not sure if it would work differently when officially hosted. For what it's worth, it looks like MuseDash uses the same format 
```py
- [Main Page](../../../../games/Muse%20Dash/info/en)
- [Options Page](../../../../games/Muse%20Dash/player-options)
```

Crazycolbster commented on worlds/openrct2/Options.py:
```diff
@@ -712,13 +712,13 @@ class RequiredUniqueRides(Range):
     option_remote = 2
     default = 0
-
+    visibility = OptionVisibility.none
```
Testing it with a fake option, that appears to be correct. I'll remove this from the options file.

Crazycolbster commented:
And with that, I've addressed all the comments. The game is essentially feature-complete on the Archipelago side. The only exception to that would be if the OpenRCT2 Devs add :wss// support in the future, which would allow me to get rid of the OpenRCT2 Client altogether. 

@Exempt-Medic, could you remove the waiting-on:author tag? Thanks!

duckboycool commented:
Something with conflict resolution seems to have had a mistake since this now apparently changes Stardew.

Crazycolbster commented:
> Something with conflict resolution seems to have had a mistake since this now apparently changes Stardew.

P*CK

I hate the stupid merge program. I just wanna make a game!

Crazycolbster commented:
Well, that either undid the merge, or it horribly borked everything. Again. P*ck Git.
